Maksim zagryadskov 3823b1pm1

midnsun opened the pull request.

midnsun committed:
some files created

midnsun committed:
MergeSort added

vector, stack, queue, merge, mergesort

midnsun committed:
Some files added

midnsun committed:
List class rethinking

midnsun committed:
myList class ready

midnsun committed:
Polinoms class development started

midnsun committed:
Addition

midnsun committed:
Code was written. Some bugs left

midnsun committed:
Bugs fixed. Testing started

I found "mutable" keyword in c++ and made some improvements to make code more clear

midnsun committed:
Started writing tests

All except polinoms are written

midnsun committed:
Project is ready

All done

midnsun commented:
@VanyokProgZ @amelabrigid Код готов к проверке

VanyokProgZ commented on include/myList.h:
```diff
@@ -0,0 +1,219 @@
+		ListNode(const T& val, ListNode* next = nullptr) : val(val), next(next) {}
+	};
+	ListNode* head;
+	mutable ListNode* lastUsed;
```
mutable вроде как не очень хорошо использовать, возможно сделать без него?

midnsun replied:
Возможно, если при каждой работе с константным списком создавать его копию для поэлементной итерации. В данной реализации подразумевается возврат указателя на текущий элемент в начало списка в начале и в конце работы со списком

VanyokProgZ commented on include/myList.h:
```diff
@@ -0,0 +1,219 @@
+		if (lastUsed == head) {
+			tmp = head;
+			head = head->next;
+			lastUsed = head;
```
Может ли здесь (при lastUsed==head) head стать nullptr? - рассматривает ли этот случай задумка?

midnsun replied:
Рассматривает. В пустом списке все указатели (head и lastUsed) указывают на nullptr. Метод isEnd(), таким образом, охватывает оба случая: когда текущий элемент находится за концом списка и когда список пуст (как частный случай предыдущего случая)

VanyokProgZ reviewed (commented):
пока только лист посмотрел

VanyokProgZ commented on include/myStack.h:
```diff
@@ -0,0 +1,34 @@
+        return this->myVector<T>::operator[](--this->myVector<T>::sizevar);
+    }
+    T& top() {
+        return this->myVector<T>::operator[](this->myVector<T>::sizevar - 1);
```
(top()): А что если sizevar равен 0?

midnsun replied:
Ровно то же самое, что и если написать vector[-1]. Предполагается, что при использовании стека (как и вектора) проверка на пустоту производится внешним образом

VanyokProgZ commented on src/polinom.cpp:
```diff
@@ -0,0 +1,335 @@
+	}
+	data.toBegin();
+	newData.toBegin();
+	eps = maxValue * 1.0e-15;
```
Почему именно такой eps?

VanyokProgZ commented:
Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)

midnsun commented:
> Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)

Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?

VanyokProgZ commented:
> > Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)
> 
> Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?

По компаратору: ты просто пропушиваешь cmpMonom в мерж, функтор выполняет обычное сравнение. Есть еще cmp, но он используется в войдах по типу "testList", т.е в конечном продукте по факту используется только один метод сравнения описанный в cmpMonom.
По оператору: у тебя есть блоки кода по смыслу отличающиеся только переменной (ну и множителем), можно вынести в отдельную функцию (и увеличить читаемость кода) 
![image](https://github.com/user-attachments/assets/a51f071d-9afa-4380-aea9-90d04de3305c)
Ну и еще у тебя используется как твой parser, так и std::stoi()

Upd: Это не влияет особо, только на читаемость

midnsun commented:
> > > Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)
> > 
> > 
> > Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?
> 
> По компаратору: ты просто пропушиваешь cmpMonom в мерж, функтор выполняет обычное сравнение. Есть еще cmp, но он используется в войдах по типу "testList", т.е в конечном продукте по факту используется только один метод сравнения описанный в cmpMonom. По оператору: у тебя есть блоки кода по смыслу отличающиеся только переменной (ну и множителем), можно вынести в отдельную функцию (и увеличить читаемость кода) ![image](https://private-user-images.githubusercontent.com/89320329/407897491-a51f071d-9afa-4380-aea9-90d04de3305c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxODMwNzUsIm5iZiI6MTczODE4Mjc3NSwicGF0aCI6Ii84OTMyMDMyOS80MDc4OTc0OTEtYTUxZjA3MWQtOWFmYS00MzgwLWFlYTktOTBkMDRkZTMzMDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI5VDIwMzI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MDJiNTdjNWQ5Mjc5YmZiOTU0MDgwNGVmZGYwM2YwOGQ0NDU4NGEyMTdhNWVhY2Y5M2MxMmNkMTExZGY3YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6mb2aGlxXywF6Ky-zm9HIAss06Dgl4zwjN-q5e-ijg) Ну и еще у тебя используется как твой parser, так и std::stoi()

По поводу parser и stoi согласен, поменяю на parser. По поводу operator>>: отдельная функция в 3 строки это неплохо, лично мне увеличение количества функций в данной ситуации снизило бы читаемость. Да, тут одинаковый код скопирован дважды - перед началом цикла и в его теле, но сам код не представляет из себя какой то сложный алгоритм. 

По поводу компаратора: я старался писать разные классы абстрактно друг от друга, то есть чтобы vector не знал, что будет наследован stack, а stack - что будет наследован queue. Аналогично и list не должен знать, что будет агрегирован классом polinom, а конкретно - что его шаблоным типом станет пара чисел int и double. Соответственно, выполнять сортировку для произвольных типов данных можно двумя способами: в надежде, что будет перегружен operator< для него или принимать функтор-comparator. Второй вариант предпочтительнее, поскольку operator< не предполагает возможность пересортировки элементов по другому признаку (например, в паре <int, double> сначала по int, потом по double)

VanyokProgZ commented:
> > > > Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)
> > > 
> > > 
> > > Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?
> > 
> > 
> > По компаратору: ты просто пропушиваешь cmpMonom в мерж, функтор выполняет обычное сравнение. Есть еще cmp, но он используется в войдах по типу "testList", т.е в конечном продукте по факту используется только один метод сравнения описанный в cmpMonom. По оператору: у тебя есть блоки кода по смыслу отличающиеся только переменной (ну и множителем), можно вынести в отдельную функцию (и увеличить читаемость кода) ![image](https://private-user-images.githubusercontent.com/89320329/407897491-a51f071d-9afa-4380-aea9-90d04de3305c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxODMwNzUsIm5iZiI6MTczODE4Mjc3NSwicGF0aCI6Ii84OTMyMDMyOS80MDc4OTc0OTEtYTUxZjA3MWQtOWFmYS00MzgwLWFlYTktOTBkMDRkZTMzMDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI5VDIwMzI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MDJiNTdjNWQ5Mjc5YmZiOTU0MDgwNGVmZGYwM2YwOGQ0NDU4NGEyMTdhNWVhY2Y5M2MxMmNkMTExZGY3YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6mb2aGlxXywF6Ky-zm9HIAss06Dgl4zwjN-q5e-ijg) Ну и еще у тебя используется как твой parser, так и std::stoi()
> 
> По поводу parser и stoi согласен, поменяю на parser. По поводу operator>>: отдельная функция в 3 строки это неплохо, лично мне увеличение количества функций в данной ситуации снизило бы читаемость. Да, тут одинаковый код скопирован дважды - перед началом цикла и в его теле, но сам код не представляет из себя какой то сложный алгоритм.
> 
> По поводу компаратора: я старался писать разные классы абстрактно друг от друга, то есть чтобы vector не знал, что будет наследован stack, а stack - что будет наследован queue. Аналогично и list не должен знать, что будет агрегирован классом polinom, а конкретно - что его шаблоным типом станет пара чисел int и double. Соответственно, выполнять сортировку для произвольных типов данных можно двумя способами: в надежде, что будет перегружен operator< для него или принимать функтор-comparator. Второй вариант предпочтительнее, поскольку operator< не предполагает возможность пересортировки элементов по другому признаку (например, в паре <int, double> сначала по int, потом по double)

В таком случае согласен

midnsun committed:
Changed std::stoi to parser

midnsun committed:
Cleared code

midnsun commented:
> > > > > Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)
> > > > 
> > > > 
> > > > Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?
> > > 
> > > 
> > > По компаратору: ты просто пропушиваешь cmpMonom в мерж, функтор выполняет обычное сравнение. Есть еще cmp, но он используется в войдах по типу "testList", т.е в конечном продукте по факту используется только один метод сравнения описанный в cmpMonom. По оператору: у тебя есть блоки кода по смыслу отличающиеся только переменной (ну и множителем), можно вынести в отдельную функцию (и увеличить читаемость кода) ![image](https://private-user-images.githubusercontent.com/89320329/407897491-a51f071d-9afa-4380-aea9-90d04de3305c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxODMwNzUsIm5iZiI6MTczODE4Mjc3NSwicGF0aCI6Ii84OTMyMDMyOS80MDc4OTc0OTEtYTUxZjA3MWQtOWFmYS00MzgwLWFlYTktOTBkMDRkZTMzMDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI5VDIwMzI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MDJiNTdjNWQ5Mjc5YmZiOTU0MDgwNGVmZGYwM2YwOGQ0NDU4NGEyMTdhNWVhY2Y5M2MxMmNkMTExZGY3YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6mb2aGlxXywF6Ky-zm9HIAss06Dgl4zwjN-q5e-ijg) Ну и еще у тебя используется как твой parser, так и std::stoi()
> > 
> > 
> > По поводу parser и stoi согласен, поменяю на parser. По поводу operator>>: отдельная функция в 3 строки это неплохо, лично мне увеличение количества функций в данной ситуации снизило бы читаемость. Да, тут одинаковый код скопирован дважды - перед началом цикла и в его теле, но сам код не представляет из себя какой то сложный алгоритм.
> > По поводу компаратора: я старался писать разные классы абстрактно друг от друга, то есть чтобы vector не знал, что будет наследован stack, а stack - что будет наследован queue. Аналогично и list не должен знать, что будет агрегирован классом polinom, а конкретно - что его шаблоным типом станет пара чисел int и double. Соответственно, выполнять сортировку для произвольных типов данных можно двумя способами: в надежде, что будет перегружен operator< для него или принимать функтор-comparator. Второй вариант предпочтительнее, поскольку operator< не предполагает возможность пересортировки элементов по другому признаку (например, в паре <int, double> сначала по int, потом по double)
> 
> В таком случае согласен

Почему не написал что у меня в коде куча закомментированного отладочного кода? Если бы не было ошибок, там бы так и осталось лишних функций в main

VanyokProgZ commented:
> > > > > > Проверил, вроде здраво, единственное можно было без comparator в мерже и оператор >> в полиноме слишком нагруженный (не хватает как будто разделения)
> > > > > 
> > > > > 
> > > > > Как ты предлагаешь избавиться от функтора comparator? И на какие функции разделить operator>>, если она вся парсит полином?
> > > > 
> > > > 
> > > > По компаратору: ты просто пропушиваешь cmpMonom в мерж, функтор выполняет обычное сравнение. Есть еще cmp, но он используется в войдах по типу "testList", т.е в конечном продукте по факту используется только один метод сравнения описанный в cmpMonom. По оператору: у тебя есть блоки кода по смыслу отличающиеся только переменной (ну и множителем), можно вынести в отдельную функцию (и увеличить читаемость кода) ![image](https://private-user-images.githubusercontent.com/89320329/407897491-a51f071d-9afa-4380-aea9-90d04de3305c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxODMwNzUsIm5iZiI6MTczODE4Mjc3NSwicGF0aCI6Ii84OTMyMDMyOS80MDc4OTc0OTEtYTUxZjA3MWQtOWFmYS00MzgwLWFlYTktOTBkMDRkZTMzMDVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI5VDIwMzI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MDJiNTdjNWQ5Mjc5YmZiOTU0MDgwNGVmZGYwM2YwOGQ0NDU4NGEyMTdhNWVhY2Y5M2MxMmNkMTExZGY3YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h6mb2aGlxXywF6Ky-zm9HIAss06Dgl4zwjN-q5e-ijg) Ну и еще у тебя используется как твой parser, так и std::stoi()
> > > 
> > > 
> > > По поводу parser и stoi согласен, поменяю на parser. По поводу operator>>: отдельная функция в 3 строки это неплохо, лично мне увеличение количества функций в данной ситуации снизило бы читаемость. Да, тут одинаковый код скопирован дважды - перед началом цикла и в его теле, но сам код не представляет из себя какой то сложный алгоритм.
> > > По поводу компаратора: я старался писать разные классы абстрактно друг от друга, то есть чтобы vector не знал, что будет наследован stack, а stack - что будет наследован queue. Аналогично и list не должен знать, что будет агрегирован классом polinom, а конкретно - что его шаблоным типом станет пара чисел int и double. Соответственно, выполнять сортировку для произвольных типов данных можно двумя способами: в надежде, что будет перегружен operator< для него или принимать функтор-comparator. Второй вариант предпочтительнее, поскольку operator< не предполагает возможность пересортировки элементов по другому признаку (например, в паре <int, double> сначала по int, потом по double)
> > 
> > 
> > В таком случае согласен
> 
> Почему не написал что у меня в коде куча закомментированного отладочного кода? Если бы не было ошибок, там бы так и осталось лишних функций в main

Так смысл придираться к комментам?

midnsun committed:
Polinom addition improvement

Made a linear merge algorithm instead of mergesort algorithm in the begining of the addition function

VanyokProgZ commented on src/polinom.cpp:
```diff
@@ -0,0 +1,351 @@
+polinom::polinom(const polinom& p) : maxPower(MAX_POWER), data(p.data) {}
+
+polinom::polinom(polinom&& p) : maxPower(MAX_POWER), data(p.data) {}
+
```
Для data тоже надо вызывать move конструктор
*сейчас если я правильно помню у тебя вызовется обычный конструктор копирования у листа

VanyokProgZ commented on src/polinom.cpp:
```diff
@@ -0,0 +1,351 @@
+}
+
+polinom& polinom::operator=(polinom&& p) {
+	data = p.data;
```
то же самое что и с move конструктором

midnsun committed:
Move functions fixed

VanyokProgZ commented:
Ошибки слишком хорошо спрятаны. Больше не нашел

midnsun commented:
@ValentinV95 Лабораторная готова

ValentinV95 reviewed (commented):
Лабораторная принята
